text/0000-ssr-rehydration.md

```diff
@@ -0,0 +1,299 @@
+
+<details open>
+  <summary><b>Example:</b></summary>
+
+An LWC-based framework could provide a new `LightningServerElement` class. Its behavior and underlying implementation could be switched out at build time, as a function of the intended runtime environment.
+
+On the server, a `LightningServerElement` could behave identically to a `LightningElement` with one exception: it would attach a `data-ssr-preserve` attribute to its root node.
+
+For a client build, a Babel plugin could identify any class declaration that extends `LightningServerElement` and replace the entire declaration with `NoopElement`.  That would look something like the following:
+
+**Original Source:**
+
+```javascript
+import { api } from 'lwc';
+import { LightningServerElement } from 'some-framework';
+import { someSideEffect } from './my-app';
+
+export default class Example extends LightningServerElement {
+    @api message = 'Hello World!';
+
+    onClick () {
+      someSideEffect();
+    }
+}
+```
+
+**Server Build:**
+
+```javascript
+import { api } from 'lwc';
+import { LightningServerElement } from 'some-framework';
+import { someSideEffect } from './my-app';
+
+export default class Example extends LightningServerElement {
+    @api message = 'Hello World!';
+
+    // LightningServerElement hooks into component lifecycle:
+    //   connectedCallback () {
+    //    this.setAttribute('data-ssr-preserve', true);
+    //   }
+
+    onClick () {
+      someSideEffect();
+    }
+}
+```
+
+**SSR HTML:**
+
+```html
+<!-- ... -->
+  <x-example data-ssr-preserve>
+    <template shadowroot="open">
+      Hello World!
+    </template>
+  </x-example>
+<!-- ... -->
+```
+
+**Client Build:**
+
+```javascript
+import { api } from 'lwc';
+import { NoopElement } from 'some-framework';
+import { someSideEffect } from './my-app'; // <-- dead-code, removed with minification
+
+// The entire class declaration is replaced.
+export default class Example extends NoopElement {}
+```
+
+</details>
+
+### SSR Placeholders
+
+Implementation of placeholder components, as described in [Exception A to the happy path](#the-happy-path) should be practicable without changes to LWC, so long as unexpected subtree mismatches (Exception C) are dealt with properly.
+
+Implementation in user-land is possible with simple runtime environment detection:
+
+<details open>
+  <summary>Example:</summary>
+
+```javascript
+import { LightningElement } from 'lwc';
+import { isNodeEnv } from './my-utils';
+import tmplServerPlaceholder from './templateServer.html';
+import tmplClient from './templateClient.html';
+
+export default class ElementWithSSRPlaceholder extends LightningElement {
+    render() {
+        return this.isNodeEnv ? tmplServerPlaceholder : tmplClient;

```

Comment:
How does this pattern address exception C? By using two different templates, there is a high chance the engine will be running in an unexpected mismatch.

For SSR placeholder components, I would envisioning the following pattern:

```js
import { LightningElement } from 'lwc';

export default class Placeholder extends LightningElement {
    showPlaceholder = true;

    connectedCallback() {
        if (!process.env.SSR) {
            setTimeout(() => this.showPlaceholder = false);
        }
    }
}
```

```html
<template>
    <template if:false={showPlaceholder}>
        <x-client-side-only></x-client-side-only>
    </template>

    <template if:true={showPlaceholder}>
        I am SSR placeholder
    </template>
</template>
```

By having the `showPlaceholder` property set to true by default, the placeholder will always be rendered on the server and the client. This avoids the unexpected mismatch during rehydration. After the component is rehydrated, the `showPlaceholder` property is set false in the following macro task to finally render the `<x-client-side-only>` element.

Reply:
I will re-word this. I wasn't intending to say that this approach solves exception C. I was saying that if we solve for exception C, implementing placeholder components (exception A) becomes straightforward.